mock/src/test_ctx.rs

```diff
@@ -0,0 +1,127 @@
+//! Mock types and functions to generate Test enviroments for ZKEVM tests
+
+use crate::{MockAccount, MockBlock, MockTransaction};
+use eth_types::{
+    geth_types::{Account, BlockConstants, GethData},
+    Block, Error, GethExecTrace, Transaction, Word,
+};
+use external_tracer::{trace, TraceConfig};
+use itertools::Itertools;
+
+/// TestContext is a type that contains all the information from a block
+/// required to build the circuit inputs.
+#[derive(Debug)]
+pub struct TestContext<const NACC: usize, const NTX: usize> {
+    /// chain id
+    pub chain_id: Word,
+    /// Account list
+    pub accounts: [Account; NACC],
+    /// history hashes contains most recent 256 block hashes in history, where
+    /// the lastest one is at history_hashes[history_hashes.len() - 1].
+    pub history_hashes: Vec<Word>,
+    /// Block from geth
+    pub eth_block: eth_types::Block<eth_types::Transaction>,
+    /// Execution Trace from geth
+    pub geth_traces: [eth_types::GethExecTrace; NTX],
+}
+
+impl<const NACC: usize, const NTX: usize> From<TestContext<NACC, NTX>> for GethData {
+    fn from(ctx: TestContext<NACC, NTX>) -> GethData {
+        GethData {
+            chain_id: ctx.chain_id,
+            history_hashes: ctx.history_hashes,
+            eth_block: ctx.eth_block,
+            geth_traces: ctx.geth_traces.to_vec(),
+            accounts: ctx.accounts.into(),
+        }
+    }
+}
+
+impl<const NACC: usize, const NTX: usize> TestContext<NACC, NTX> {
+    pub fn new<FAcc, FTx, Fb>(
+        history_hashes: Option<Vec<Word>>,
+        acc_fns: FAcc,
+        func_tx: FTx,
+        func_block: Fb,
+    ) -> Result<Self, Error>
+    where
+        FTx: FnOnce(Vec<&mut MockTransaction>, [MockAccount; NACC]),
+        Fb: FnOnce(&mut MockBlock, Vec<MockTransaction>) -> &mut MockBlock,
+        FAcc: FnOnce([&mut MockAccount; NACC]) -> [&mut MockAccount; NACC],
```

Comment:
The definition of the `func_block` type requires returning the `MockBlock`, and the definition of the `acc_fns` require returning the `MockAccount`s.
On the other hand, the type of `func_tx` doesn't require a return value.
I wonder if `FAcc` could be changed to not require a return value?  This would eliminate a line from the closure every time:
From this:
```rust
    |accs| {
        accs[0]
            .address(address!("0x0000000000000000000000000000000000000000"))
            .code(code_a);
        accs[1]
            .address(address!("0x000000000000000000000000000000000cafe001"))
            .code(code_b);
        accs
    },
```
To this:
```rust
    |accs| {
        accs[0]
            .address(address!("0x0000000000000000000000000000000000000000"))
            .code(code_a);
        accs[1]
            .address(address!("0x000000000000000000000000000000000cafe001"))
            .code(code_b);
    },
```

I think the `Fb` trait is already fine because the closure is just one line already:
```rust
    |block, _tx| block.number(0xcafeu64),
```

Reply:
It makes complete sense. I indeed went for this option for the Tx but forgot to update also the block functions..

Thanks for the catch!!! :)